### PR TITLE
faketikv: add client

### PR DESCRIPTION
**Pull request by nolouch**
add GRPC client.

### CODE REVIEW

File: pkg/faketikv/client.go [hunk 1]
```diff
@@ -25,5 +39,280 @@ type Client interface {
...
+	)
+	for {
+		if stream == nil {
+			ctx, cancel = context.WithCancel(c.ctx)
+			stream, err = c.pdClient.RegionHeartbeat(ctx)
+			if err != nil {
+				log.Errorf("[store %d][pd] create region heartbeat stream error: %v", c.tag, err)
+				cancel()
+				select {
+				case <-time.After(time.Second):
```
Comment (Connor1996) [inline pkg/faketikv/client.go]: no need for select, maybe use time.Sleep instead?

File: pkg/faketikv/client.go [hunk 2]
```diff
@@ -25,5 +39,280 @@ type Client interface {
...
+			log.Infof("[store %d][pd] heartbeat stream get error: %s times: %d", c.tag, err, i)
+			cancel()
+		case <-c.ctx.Done():
+			return
+		}
+		wg.Wait()
+	}
+}
+
+func (c *client) reciveRegionHeartbeat(ctx context.Context, stream pdpb.PD_RegionHeartbeatClient, errCh chan error, wg *sync.WaitGroup) {
```
Comment (Connor1996) [inline pkg/faketikv/client.go]: recive -> receive?

File: pkg/faketikv/client.go [hunk 3]
```diff
@@ -25,5 +39,280 @@ type Client interface {
...
+	maxInitClusterRetries = 100
+)
+
+var (
+	// errFailInitClusterID is returned when failed to load clusterID from all supplied PD addresses.
+	errFailInitClusterID = errors.New("[pd] failed to get cluster id")
+	// errClosing is returned when request is canceled when client is closing.
+	errClosing = errors.New("[pd] closing")
+	// errTSOLength is returned when the number of response timestamps is inconsistent with request.
+	errTSOLength = errors.New("[pd] tso length in rpc response is incorrect")
```
Comment (disksing) [inline pkg/faketikv/client.go]: unused?

File: pkg/faketikv/client.go [hunk 4]
```diff
@@ -25,5 +39,280 @@ type Client interface {
...
+const (
+	pdTimeout             = time.Second
+	maxInitClusterRetries = 100
+)
+
+var (
+	// errFailInitClusterID is returned when failed to load clusterID from all supplied PD addresses.
+	errFailInitClusterID = errors.New("[pd] failed to get cluster id")
+	// errClosing is returned when request is canceled when client is closing.
+	errClosing = errors.New("[pd] closing")
```
Comment (disksing) [inline pkg/faketikv/client.go]: unused?

File: pkg/faketikv/client.go [hunk 5]
```diff
@@ -25,5 +39,277 @@ type Client interface {
...
+
+func (c *client) createHeartbeatStream() (pdpb.PD_RegionHeartbeatClient, context.Context, context.CancelFunc) {
+	var (
+		stream pdpb.PD_RegionHeartbeatClient
+		err    error
+		cancel context.CancelFunc
+		ctx    context.Context
+	)
+	for {
+		if stream == nil {
```
Comment (disksing) [inline pkg/faketikv/client.go]: I think it is always nil.

File: pkg/faketikv/client.go [hunk 6]
```diff
@@ -147,7 +147,7 @@ func (c *client) createHeartbeatStream() (pdpb.PD_RegionHeartbeatClient, context
 					continue
 				case <-c.ctx.Done():
 					log.Info("cancel create stream loop")
-					return nil, ctx, context
+					return nil, ctx, cacel
```
Comment (Connor1996) [inline pkg/faketikv/client.go]: cacel -> cancel

**Comment by disksing** [comment]
LGTM. PTAL @Connor1996

File: pkg/faketikv/client.go [hunk 7]
```diff
@@ -25,5 +39,273 @@ type Client interface {
...
+	return nil
+}
+
+func (c *client) requestHeader() *pdpb.RequestHeader {
+	return &pdpb.RequestHeader{
+		ClusterId: c.clusterID,
+	}
+}
+
+func addrsToUrls(addr string) string {
```
Comment (Connor1996) [inline pkg/faketikv/client.go]: no need for multiple addrs? if so addrsToUrls -> addrToUrl

Comment (siddontang) [reply pkg/faketikv/client.go]: it is for test, seem only one PD is enough.

File: pkg/faketikv/client.go [hunk 8]
```diff
@@ -25,5 +39,273 @@ type Client interface {
...
+
+	wg     sync.WaitGroup
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
 // NewClient creates a PD client.
-func NewClient(pdAddr string) (Client, <-chan *pdpb.RegionHeartbeatResponse, error) {}
+func NewClient(pdAddr string, tag uint64) (Client, <-chan *pdpb.RegionHeartbeatResponse, error) {
+	log.Infof("[store %d][pd] create pd client with endpoints %v", tag, pdAddr)
```
Comment (siddontang) [inline pkg/faketikv/client.go]: use `[store xxx]` as the tag.

File: pkg/faketikv/client.go (hunk 7 above)
Comment (siddontang) [reply pkg/faketikv/client.go]: but I don't think we need this function, forcibly using http is fine.

File: pkg/faketikv/client.go [hunk 9]
```diff
@@ -25,5 +39,273 @@ type Client interface {
...
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (c *client) Close() {
+	c.cancel()
+	c.wg.Wait()
```
Comment (siddontang) [inline pkg/faketikv/client.go]: do we need to close the connection before wait?

File: pkg/faketikv/client.go [hunk 10]
```diff
@@ -25,5 +39,273 @@ type Client interface {
...
+
+func (c *client) getMembers(ctx context.Context) (*pdpb.GetMembersResponse, error) {
+	members, err := c.pdClient().GetMembers(ctx, &pdpb.GetMembersRequest{})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return members, nil
+}
+
+func (c *client) createGRPCConn() (*grpc.ClientConn, error) {
```
Comment (siddontang) [inline pkg/faketikv/client.go]: createConn is enough.

**Comment by siddontang** [comment]
Rest LGTM

File: pkg/faketikv/client.go (hunk 9 above)
Comment (nolouch) [reply pkg/faketikv/client.go]: both are ok.

**Pull request merged**